src/front/spv.rs

```diff
@@ -2055,51 +2148,86 @@ impl<I: Iterator<Item = u32>> Parser<I> {
                     fun_inst.expect(2)?;
                     let label_id = self.next()?;
+
                     let node = self.next_block(
+                        label_id,
                         &mut fun.expressions,
                         &mut fun.local_variables,
                         &module.types,
                         &module.constants,
                         &module.global_variables,
                         &mut local_function_calls,
                     )?;
-                    // temp until the CFG is fully processed
-                    for assign in node.assignments.iter() {
-                        fun.body.push(crate::Statement::Store {
-                            pointer: assign.to,
-                            value: assign.value,
-                        });
-                    }
-                    match node.terminator {
-                        Terminator::Return { value } => {
-                            fun.body.push(crate::Statement::Return { value });
-                        }
-                        Terminator::Branch {
-                            label_id,
-                            condition,
-                        } => {
-                            let _ = (label_id, condition); //TODO
-                        }
-                    }
-                    control_flow_graph.insert(label_id, node);
+
+                    id_to_node.insert(label_id, control_flow_graph.add_node(node));
                 }
                 spirv::Op::FunctionEnd => {
                     fun_inst.expect(1)?;
                     break;
                 }
-                _ => return Err(Error::UnsupportedInstruction(self.state, fun_inst.op)),
+                _ => panic!("SHOULD NOT HAPPEN"),
+            }
+        }
+
+        // Create Edges.
+        let mut edges = Vec::new();
+        for node in control_flow_graph.node_weights_mut() {
+            let source_node_index = id_to_node[&node.id];
+
+            match node.terminator {
+                Terminator::Branch { target_id } => {
+                    let target_node_index = id_to_node[&target_id];
+
+                    edges.push((
+                        source_node_index,
+                        target_node_index,
+                        ControlFlowEdge::Forward,
+                    ));
+                }
+                Terminator::BranchConditional {
+                    true_id, false_id, ..
+                } => {
+                    let true_node_index = id_to_node[&true_id];
+                    let false_node_index = id_to_node[&false_id];
+
+                    edges.push((source_node_index, true_node_index, ControlFlowEdge::IfTrue));
+                    edges.push((
+                        source_node_index,
+                        false_node_index,
+                        ControlFlowEdge::IfFalse,
+                    ));
+                }
+                _ => {}
             }
         }
+        for edge in edges {
```

Comment:
can we have this written as `for (src, dst, mode) in edges` or something like this?

Reply:
Of course. Will fix